Framework/HistogramData/src/HistogramRebin.cpp

```diff
@@ -0,0 +1,157 @@
+#include "MantidHistogramData/HistogramRebin.h"
+#include "MantidHistogramData/BinEdges.h"
+#include "MantidHistogramData/Histogram.h"
+#include <algorithm>
+#include <numeric>
+
+using Mantid::HistogramData::Histogram;
+using Mantid::HistogramData::BinEdges;
+using Mantid::HistogramData::Counts;
+using Mantid::HistogramData::CountStandardDeviations;
+using Mantid::HistogramData::CountVariances;
+using Mantid::HistogramData::Frequencies;
+using Mantid::HistogramData::FrequencyStandardDeviations;
+using Mantid::HistogramData::FrequencyVariances;
+
+namespace {
+Histogram rebinCounts(const Histogram &input, const BinEdges &binEdges) {
+  if (input.yMode() != Histogram::YMode::Counts ||
+      input.xMode() != Histogram::XMode::BinEdges) {
+    throw std::runtime_error(
+        "Histogram  XMode should be BinEdges and YMode should be Counts.");
+  }
```

Comment:
I don't think this needs to be checked here again, since this function is only visible within this file...